staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -214,8 +214,9 @@ func (h *peerProxyHandler) resolveServingLocation(apiservers *sync.Map) (bool, [
 		return true
 	})
 
-	// reset err if there was atleast one valid peer server found.
-	if len(peerServerEndpoints) > 0 {
+	// reset err if request can be served locally or
+	// if there was atleast one valid peer server found.
+	if locallyServiceable || len(peerServerEndpoints) > 0 {
```

Comment:
this seems a clear bug

Reply:
however, the test fails in 

> // list jobs using ServerB which it should proxy to ServerC and get back valid response

so, this can not be the fix IIUIC

Reply:
hmm, not a bug, the respErr is not only only if the hostPort information is not parsed

Reply:
ok, this is correct, if we find a locallyServiceable we just shortcut and exit , @richabanker please see this patch, I also notice the unit test was not enabling the feature gates for the local testcase

```
diff --git a/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go b/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go
index 8e236a7505c..71d4e558b25 100644
--- a/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go
+++ b/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go
@@ -199,7 +199,8 @@ func (h *peerProxyHandler) resolveServingLocation(apiservers *sync.Map) (bool, [
                apiserverKey := key.(string)
                if apiserverKey == h.serverId {
                        locallyServiceable = true
-                       // stop iteration
+                       // stop iteration and clean any possible errors
+                       respErr = nil
                        return false
                }
 
diff --git a/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler_test.go b/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler_test.go
index f9c9aec9e26..1a7a36b6da4 100644
--- a/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler_test.go
@@ -146,6 +146,67 @@ func TestPeerProxy(t *testing.T) {
                                        Resource: "bar"},
                                serverIDs: []string{localServerID}},
                },
+               {
+                       desc:                 "200 if locally serviceable feature enable",
+                       requestPath:          "/api/foo/bar",
+                       expectedStatus:       http.StatusOK,
+                       informerFinishedSync: true,
+                       svdata: FakeSVMapData{
+                               gvr: schema.GroupVersionResource{
+                                       Group:    "core",
+                                       Version:  "foo",
+                                       Resource: "bar"},
+                               serverIDs: []string{localServerID}},
+                       reconcilerConfig: reconciler{
+                               do: true,
+                       },
+               },
+               {
+                       desc:                 "200 if locally serviceable and peer with error",
+                       requestPath:          "/api/foo/bar",
+                       expectedStatus:       http.StatusOK,
+                       informerFinishedSync: true,
+                       svdata: FakeSVMapData{
+                               gvr: schema.GroupVersionResource{
+                                       Group:    "core",
+                                       Version:  "foo",
+                                       Resource: "bar"},
+                               serverIDs: []string{localServerID}},
+                       reconcilerConfig: reconciler{
+                               do: true,
+                               servers: []server{
+                                       {
+                                               publicIP: "1.2.3.4",
+                                               serverID: localServerID,
+                                       },
+                               },
+                       },
+               },
+               {
+                       desc:                 "200 if locally serviceable and peer with error",
+                       requestPath:          "/api/foo/bar",
+                       expectedStatus:       http.StatusOK,
+                       informerFinishedSync: true,
+                       svdata: FakeSVMapData{
+                               gvr: schema.GroupVersionResource{
+                                       Group:    "core",
+                                       Version:  "foo",
+                                       Resource: "bar"},
+                               serverIDs: []string{localServerID, remoteServerID}},
+                       reconcilerConfig: reconciler{
+                               do: true,
+                               servers: []server{
+                                       {
+                                               publicIP: "1.2.3.4",
+                                               serverID: localServerID,
+                                       },
+                                       {
+                                               publicIP: "1.2.wrong.4",
+                                               serverID: remoteServerID,
+                                       },
+                               },
+                       },
+               },
                {
                        desc:                 "503 unreachable peer bind address",
                        requestPath:          "/api/foo/bar",
@@ -285,7 +346,10 @@ func TestPeerProxy(t *testing.T) {
                        }
                        req.Header.Set(PeerProxiedHeader, tt.peerproxiedHeader)
 
-                       resp, _ := requestGetter(req)
+                       resp, err := requestGetter(req)
+                       if err != nil {
+                               t.Fatalf("unexpected error trying to get the request: %v", err)
+                       }
 
                        // compare response
                        assert.Equal(t, tt.expectedStatus, resp.StatusCode)
```

Reply:
hmm, I don\t seem to trigger the error with the unit test when `respErr` is not nil

Reply:
> hmm, I don\t seem to trigger the error with the unit test when `respErr` is not nil

I dont think the handler for peerproxy is returning the error currently, it [observes and logs the error](https://github.com/kubernetes/kubernetes/blob/d19e92ec56927bfccce39ea7f36c601f2ba31424/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go#L155-L160) and returns a 503 currently. 

The 503 being returned above is already being tested under the case "503 if all peers had invalid host:port info"